conan/internal/util/semaphore.py

```diff
@@ -0,0 +1,161 @@
+""" The semaphore module provides inter-process locking mechanisms to ensure Conan commands can
+    run concurrently without conflicts.
+
+    It uses the fasteners library to create and manage locks across multiple processes. Thus, this
+    module is a proxy in case the project need to use a different library in the future.
+
+    A timeout is defined to prevent deadlocks, and the lock files are stored in a temporary directory
+    in the conan cache folder. Still, the timeout is configured in seconds, using the configuration
+    `core.cache.parallel:timeout` in the conan.conf file.
+    The default value is 300 seconds (5 minutes).
+
+    The fasteners have context managers, but only acquire() method has a timeout
+    parameter, that's why we needed to extend sleep method to inject the timeout check.
+"""
+import functools
+import os
+import time
+from datetime import datetime
+
+from conan.errors import ConanException
+from conan.api.output import ConanOutput
+from conan.internal.cache.cache import PkgCache
+
+
+CONAN_SEMAPHORE_TIMEOUT = 300.0  # 5 minutes
+CONAN_SEMAPHORE_LOCKFILE = "conan_semaphore.lock"
```

Comment:
Avoid ``LOCKFILE`` naming, use ``filelocks`` instead, that will help when reading and searching in the codebase